concent_api/core/payments/sci_callback.py

```diff
@@ -0,0 +1,260 @@
+    """
+    This is callback that submits the transaction to the Signing Service
+    and does not rely on having access to the private key.
+
+    Returns signed Ethereum Transaction.
+
+    The flow is:
+    1. Create a TransactionSigningRequest.
+    2. Establish a TCP connection with Middleman.
+    3. Send the TransactionSigningRequest.
+    4. Read the response and close the connection.
+    5a. If the response is a valid SignedTransaction, copy the signature to the transaction object passed to the
+        callback by SCI and return.
+    5b. Otherwise handle errors by raising an exception and interrupting the callback.
+
+    """
+    assert isinstance(transaction, Transaction)
+
+    # Create a TransactionSigningRequest.
+    transaction_signing_request = create_transaction_signing_request(transaction)
+
+    # Generate request ID.
+    request_id = RequestIDGenerator.generate_request_id()
+
+    # Create a MiddleMan Protocol Golem Message Frame.
+    middleman_message = GolemMessageFrame(
+        payload=transaction_signing_request,
+        request_id=request_id,
+    )
+
+    # Send Frame to MiddleMan through MiddleMan Protocol and receive response.
+    raw_response = send_request_to_middleman(middleman_message)
+
+    # Deserialize received Frame and its payload and handle related errors.
+    signed_transaction = deserialize_response_and_handle_errors(raw_response, request_id)
+
+    assert isinstance(signed_transaction, SignedTransaction)
+
+    # Verify received SignedTransaction signature and handle related errors.
+    verify_data_and_signature(signed_transaction, transaction)
+
+    # If the response is a valid SignedTransaction, copy Ethereum Transaction object into new one and
+    # add SignedTransaction data.
+    return copy_transaction_and_signature(signed_transaction, transaction)
+
+
+def create_transaction_signing_request(transaction: Transaction) -> TransactionSigningRequest:
+    """ Create TransactionSigningRequest message from given Ethereum Transaction object. """
+
+    assert isinstance(transaction, Transaction)
+
+    transaction_signing_request = TransactionSigningRequest(
+        nonce    = transaction.nonce,
+        gasprice = transaction.gasprice,
+        startgas = transaction.startgas,
+        to       = transaction.to,
+        value    = transaction.value,
+        data     = transaction.data,
+    )
+
+    setattr(
+        transaction_signing_request,
+        'from',
+        generate_ethereum_address_from_ethereum_public_key_bytes(settings.CONCENT_ETHEREUM_PUBLIC_KEY),
+    )
+
+    return transaction_signing_request
+
+
+def send_request_to_middleman(middleman_message: GolemMessageFrame) -> bytes:
+    """
+    Opens socket connection to Middleman, sends Frame to MiddleMan through MiddleMan Protocol and receive response.
+
+    Returns raw Frame as bytes.
+    """
+
+    assert isinstance(middleman_message, GolemMessageFrame)
+
+    # Establish a TCP connection with Middleman.
```

Comment:
this comment seems unnecessary - the code is clear about itself ;)

Reply:
:+1: